In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Disable warnings
import warnings
warnings.filterwarnings('ignore')

In [5]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  
coin_id                                   
bitcoin                         37.51761  
ethereum                       101.96023  
tether                           0.01954  
ripple                         -16.60193  
bitcoin-cash                    14.49384

In [4]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [5]:
df_market_data.shape

(41, 7)

In [6]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [6]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [7]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data[[
       'price_change_percentage_24h', 
       'price_change_percentage_7d',
       'price_change_percentage_14d', 
       'price_change_percentage_30d',
       'price_change_percentage_60d', 
       'price_change_percentage_200d',
       'price_change_percentage_1y']])


In [8]:
# Create a DataFrame with the scaled data
df_crypto_scaled = pd.DataFrame(crypto_scaled,
                                columns=['price_change_percentage_24h', 
                                         'price_change_percentage_7d',
                                         'price_change_percentage_14d', 
                                         'price_change_percentage_30d',
                                         'price_change_percentage_60d', 
                                         'price_change_percentage_200d',
                                         'price_change_percentage_1y'])

# Copy the crypto names from the original DataFrame
df_crypto_scaled['coin_id'] = df_market_data.index

# Set the coin_id column as index
df_crypto_scaled = df_crypto_scaled.set_index('coin_id')

# Display the scaled DataFrame
df_crypto_scaled.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

In [9]:
df_crypto_scaled.shape

(41, 7)

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [10]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [11]:
# Create an empty list to store the inertia values
inertia=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_crypto_scaled)
    inertia.append(k_model.inertia_)
    

In [12]:
# Create a dictionary with the data to plot the Elbow curve
k_dict = {'k':k, 'inertia':inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_crypto_elbow = pd.DataFrame(k_dict)

df_crypto_elbow.head()

k     inertia
0  1  287.000000
1  2  212.123342
2  3  165.136752
3  4   79.022435
4  5   66.413051

In [47]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
crypto_elbow_plot = df_crypto_elbow.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve',
    xticks=k
)

crypto_elbow_plot


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [17]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [18]:
# Fit the K-Means model using the scaled DataFrame
model.fit(df_crypto_scaled)

KMeans(n_clusters=4, random_state=1)

In [19]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
crypto_clusters = model.predict(df_crypto_scaled)

# Print the resulting array of cluster values.
crypto_clusters[:10]

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0])

In [20]:
# Create a copy of the scaled DataFrame
df_crypto_scaled_predictions = df_crypto_scaled.copy()

In [21]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_crypto_scaled_predictions['CryptoCluster'] = crypto_clusters

# Display the copy of the scaled DataFrame
df_crypto_scaled_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  CryptoCluster  
coin_id                                                  
bitcoin                        -0.251637              2  
ethereum                       -0.199352              2  
tether                         -0.282061              0  
ripple                         -0.295546              0  
bitcoin-cash                   -0.270317              2

In [43]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
crypto_scaled_predictions_scatter_plot = df_crypto_scaled_predictions.hvplot.scatter(
    x='price_change_percentage_24h',
    y='price_change_percentage_7d',
    by='CryptoCluster',
    hover_cols='coin_id',
    title='Crypto Scaled Scatter Plot'  
)

crypto_scaled_predictions_scatter_plot

:NdOverlay   [CryptoCluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [23]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [24]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
crypto_pca = pca.fit_transform(df_crypto_scaled)

# View the scaled PCA data
crypto_pca[:10]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292]])

In [25]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 
37.2% + 34.7% + 17.7% = 89.6%

In [26]:
# Create a new DataFrame with the PCA data.
df_crypto_pca = pd.DataFrame(crypto_pca,
                             columns=['PC1','PC2','PC3'])

# Copy the crypto names from the original scaled DataFrame
df_crypto_pca['coin_id'] = df_market_data.index

# Set the coin_id column as index
df_crypto_pca = df_crypto_pca.set_index('coin_id')

# Display the scaled PCA DataFrame
df_crypto_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [27]:
# Create a list with the number of k-values from 1 to 11
k=list(range(1,11))

In [28]:
# Create an empty list to store the inertia values
inertia=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model_pca = KMeans(n_clusters=i, random_state=1)
    k_model_pca.fit(df_crypto_pca)
    inertia.append(k_model_pca.inertia_)
       

In [29]:
# Create a dictionary with the data to plot the Elbow curve
k_dict = {'k':k, 'inertia':inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_crypto_elbow_pca = pd.DataFrame(k_dict)

df_crypto_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  182.339530
2  3  135.442408
3  4   49.665497
4  5   38.672582

In [44]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pca_crypto_elbow = df_crypto_elbow_pca.hvplot.line(
    x='k',
    y='inertia',
    title='Elbow Curve PCA',
    xticks=k
)

pca_crypto_elbow

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**
Based on the above line graph, 4 is the best value for k.

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** 
  No, both elbow line graphs show that 4 is the best value for k.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [31]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [32]:
# Fit the K-Means model using the PCA data
model.fit(df_crypto_pca)

KMeans(n_clusters=4, random_state=1)

In [33]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
crypto_clusters = model.predict(df_crypto_pca)

# Print the resulting array of cluster values.
crypto_clusters

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [34]:
# Create a copy of the scaled PCA DataFrame
df_crypto_pca_predictions = df_crypto_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_crypto_pca_predictions['CryptoCluster'] = crypto_clusters

# Display the copy of the scaled PCA DataFrame
df_crypto_pca_predictions.head()

PC1       PC2       PC3  CryptoCluster
coin_id                                                  
bitcoin      -0.600667  0.842760  0.461595              2
ethereum     -0.458261  0.458466  0.952877              2
tether       -0.433070 -0.168126 -0.641752              0
ripple       -0.471835 -0.222660 -0.479053              0
bitcoin-cash -1.157800  2.041209  1.859715              2

In [45]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
crypto_pca_predictions_scatter_plot = df_crypto_pca_predictions.hvplot.scatter(
    x='PC1',
    y='PC2',
    by='CryptoCluster',
    hover_cols='coin_id',
    title='Crypto PCA Scaled Scatter Plot'
)

crypto_pca_predictions_scatter_plot


:NdOverlay   [CryptoCluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [46]:
# Composite plot to contrast the Elbow curves
crypto_elbow_plot + pca_crypto_elbow

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [42]:
# Composite plot to contrast the clusters
crypto_scaled_predictions_scatter_plot + crypto_pca_predictions_scatter_plot

:Layout
   .NdOverlay.I  :NdOverlay   [CryptoCluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [CryptoCluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features to cluster the data, resulted in the data to be more concentrated compared to the original scatter plot.